### PR TITLE
preserve this when extracting functions

### LINKED ISSUE #39931: `this` is not preserved when extracted to function in module scope
**TypeScript Version:**  3.9.6

**Search Terms:** extract to function, this

**Code**

```ts
function test(this: any, foo: string) {
  console.log(this.name);
  console.log(foo); 
}
```

Select these two `console.log` and extract them to function in module scope.

**Expected behavior:**
`this` should be passed to `newFunction`

```ts
function test(this: any, foo: string) {
    newFunction(this, foo);
}

function newFunction(context: any, foo: string) {
    console.log(context.name);
    console.log(foo);
}
```

**Actual behavior:**
`this` in `newFunction` is `undefined` now.
```ts
function test(this: any, foo: string) {
    newFunction(foo);
}

function newFunction(foo: string) {
    console.log(this.name);
    console.log(foo);
}

```

**Playground Link:** 
Seems Refactoring is not supported on TypeScript Playground.

**Related Issues:** 

function-this

### PR DESCRIPTION
**Pull request by Zzzen**
Fixes #39931

Previously, `this` is assumed to be a reference to class. Now we can detect whether `this` is referring to a class or function. in the latter case, we need to pass `this` to the new function.

### CODE REVIEW

**Review by sandersn — CHANGES_REQUESTED** [review]
Quick suggestion about naming enum member

File: src/services/refactors/extractSymbol.ts [hunk 1]
```diff
@@ -202,10 +203,11 @@ namespace ts.refactor.extractSymbol {
         IsGenerator = 1 << 1,
         IsAsyncFunction = 1 << 2,
         UsesThis = 1 << 3,
+        IsThisReferringToFunction = 1 << 4,
```
[the remark this answers is not in the document]
Comment (Zzzen) [reply src/services/refactors/extractSymbol.ts]: 👌

File: src/services/refactors/extractSymbol.ts [hunk 2]
```diff
@@ -606,13 +628,18 @@ namespace ts.refactor.extractSymbol {
...
-            const containingClass = getContainingClass(current);
-            if (containingClass) {
-                const containingFunction = findAncestor(current, isFunctionLikeDeclaration);
-                return containingFunction
-                    ? [containingFunction, containingClass]
-                    : [containingClass];
+            if (range.facts & RangeFacts.UsesThisInFunction) {
+                // fall through
+            }
+            else {
```
Comment (jakebailey) [inline src/services/refactors/extractSymbol.ts:634]: This is a little unusual (I've only seen `// fall through` for unbroken switch case blocks), how about:

```suggestion
            if ((range.facts & RangeFacts.UsesThisInFunction) !== 0) {
```

Comment (jakebailey) [reply src/services/refactors/extractSymbol.ts:634]: Or, move the above change up into the first `if`.

**Comment by sandersn** [comment]
Do you want to keep working on this? Otherwise I'd like to close it to reduce the number of open PRs.

**Comment by Zzzen** [comment]
Yes, I will keep working on this.

File: src/services/refactors/extractSymbol.ts (hunk 2 above)
Comment (Zzzen) [reply src/services/refactors/extractSymbol.ts]: updated

**Review by jakebailey — APPROVED** [review]
This seems fine to me, no pun intended.

**Review by sandersn — APPROVED** [review]

**Pull request merged**